### PR TITLE
`FILTER` clause in `SELECT`'s `target_elem`

### LINKED ISSUE #953: sql: implement FILTER clauses for aggregate expressions
PR: https://github.com/cockroachdb/cockroach/pull/10146

### PR DESCRIPTION
**Pull request by sploiselle**
Closes #953 

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/967)

### CODE REVIEW

**Comment by jseldess** [comment]
This PR includes a bunch of commits related to the constraints copyediting you did in another PR. Is that intentional, @sploiselle?

**Comment by sploiselle** [comment]
@jseldess Just removed them.

File: select.md [hunk 1]
```diff
@@ -274,6 +274,21 @@ WHERE state_opened IN ('AZ', 'NY', 'WA');
 +--------------+
 ~~~
 
+#### Filter Columns Fed into Aggregate Functions
+
+You can perform a `FILTER (WHERE <Boolean expression>)` in the `target_elem` to filter which rows are processed by the aggregate function; those that return `FALSE` for the `FILTER` clause's Boolean expression are not fed into the aggregate function. You can also think of this as a shorthand for generating subqueries.
```
Comment (jseldess) [inline select.md]: nit: I'd say "You can use a..." instead of "You can perform a...".

**Comment by jseldess** [comment]
One other thought: Should we add a note here pointing to the [aggregate function examples](https://www.cockroachlabs.com/docs/select.html#aggregate-functions) in the `SELECT` doc? Might be a good place to bubble these up...

**Comment by sploiselle** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[select.md, line 279 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jHHa1cPzguy7eJB7U-r1-279:-K_jT50uzzEF6uQyyigM:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/d9a0761ab6ab744086137ff491157fa498250070/select.md#L279)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

nit: I'd say "You can use a..." instead of "You can perform a...".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/967)*

**Comment by andreimatei** [comment]
:lgtm: but please see comments

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[select.md, line 279 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jWik9cWH065YcImSk:-K_jWik9cWH065YcImSl:b-6vswgy) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L279)):*
> ```Markdown
> #### Filter Columns Fed into Aggregate Functions
> 
> You can use a `FILTER (WHERE <Boolean expression>)` in the `target_elem` to filter which rows are processed by the aggregate function; those that return `FALSE` for the `FILTER` clause's Boolean expression are not fed into the aggregate function. You can also think of this as a shorthand for generating subqueries.
> ```

this is not technically correct - the rows for which the expression returns NULL are also rejected. "only rows for which the filter evaluates to true ..."

---

*[select.md, line 279 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jX-P8gQfKtBR7jIV5:-K_jX-P9I_wrR9AqBEla:b-h7o81a) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L279)):*
> ```Markdown
> #### Filter Columns Fed into Aggregate Functions
> 
> You can use a `FILTER (WHERE <Boolean expression>)` in the `target_elem` to filter which rows are processed by the aggregate function; those that return `FALSE` for the `FILTER` clause's Boolean expression are not fed into the aggregate function. You can also think of this as a shorthand for generating subqueries.
> ```

not sure what you meant by "shorthand for generating subqueries". Could you explain, or strike it?

---

*[select.md, line 282 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jXuuWuk4L3No-KsTm:-K_jXuuWuk4L3No-KsTn:b-z0j9fq) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L282)):*
> ```Markdown
> 
> ~~~ sql
> > SELECT count(*) FILTER (WHERE balance > 1500) AS "balance > 1500" FROM bank.accounts;
> ```

I'd suggest putting `count(*) as unfiltered, count(*) filter (where ...)` side by side, for people to get the idea. Otherwise, the filter could just have been moved to a regular where clause. The point is that different aggregations can have different filter.

---

*[select.md, line 282 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jYGbAKABoaF6N2Hqp:-K_jYGbAKABoaF6N2Hqq:b-j170w9) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L282)):*
> ```Markdown
> 
> ~~~ sql
> > SELECT count(*) FILTER (WHERE balance > 1500) AS "balance > 1500" FROM bank.accounts;
> ```

other queries in this page don't qualify "accounts" with "bank.account"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/967#-:-K_jYc55CKUaGwQBUZLC:bdpnmll)*

**Comment by sploiselle** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[select.md, line 279 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jWik9cWH065YcImSk:-K_jdA6Bo-Xxs83u-Umm:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L279)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this is not technically correct - the rows for which the expression returns NULL are also rejected. "only rows for which the filter evaluates to true ..."
</blockquote></details>

Done.

---

*[select.md, line 279 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jX-P8gQfKtBR7jIV5:-K_jdAYdfuYmxlqXOib2:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L279)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

not sure what you meant by "shorthand for generating subqueries". Could you explain, or strike it?
</blockquote></details>

Done.

---

*[select.md, line 282 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jXuuWuk4L3No-KsTm:-K_jdAsrB7eZjwimdF1g:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L282)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd suggest putting `count(*) as unfiltered, count(*) filter (where ...)` side by side, for people to get the idea. Otherwise, the filter could just have been moved to a regular where clause. The point is that different aggregations can have different filter.
</blockquote></details>

Done.

---

*[select.md, line 282 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/967#-K_jYGbAKABoaF6N2Hqp:-K_jdBC_MhAi1kT-cOue:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5519388f7c3303923795b6077915be3581cd9b04/select.md#L282)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

other queries in this page don't qualify "accounts" with "bank.account"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/967)*

**Pull request merged**